contracts/SL2RD_V2.sol

```diff
@@ -0,0 +1,468 @@
+            deleteShareholderNode(msg.sender);
+        }
+        return true;
+    }
+
+    /// @notice Allows for ERC20 tokens to be transferred to a new address.
+    /// @dev Overrides the ERC20 version to add additional check that ensures
+    /// the recipient address is a SHARE approved wallet hash as well as
+    /// internal linked list updates. This function is designed to be
+    /// called from token holders as well as SHARE protocol registered
+    /// operators. If the caller is a SHARE protocol registered operator,
+    /// the `from` value must be the contract owner. In other words,
+    /// only the contract owner is delegating any priviledges to the
+    /// SHARE protocol registered operator. Shareholders that are not
+    /// the contract owner must call this function directly or approve
+    /// delegation to an exchange contract.
+    function transferFrom(
+        address from_,
+        address to_,
+        uint256 value_
+    ) public override returns (bool) {
+        if (_codeVerificationEnabled) {
+            require(
+                _protocol.isApprovedBuild(
+                    to_,
+                    CodeVerification.BuildType.WALLET
+                ) ||
+                    isApprovedLiquidityPoolCodeHash(
+                        CodeVerification.readCodeHash(to_)
+                    ),
+                "SHARE007"
+            );
+        }
+
+        if (
+            _shareOperatorRegistry.isOperator(msg.sender) ||
+            msg.sender == owner()
+        ) {
+            require(_testMode || from_ == owner(), "SHARE041");
+            super._approve(from_ /* owner */, msg.sender /* spender */, value_);
+        }
+
+        if (balanceOf(to_) == 0) {
+            addShareholderNode(to_);
+        }
+        super.transferFrom(from_, to_, value_);
+        if (balanceOf(from_) == 0) {
+            deleteShareholderNode(from_);
+        }
+        return true;
+    }
+
+    /// @notice Transfers shares from public allocation (e.g. owned by the contract owner)
+    /// to a new shareholder.
+    /// @param to_ The address of the shareholder to transfer shares to.
+    /// @param value_ The number of shares to transfer.
+    function transferPublicShares(
+        address to_,
+        uint256 value_
+    ) public onlyOwnerOrOperator nonReentrant {
+        require(
+            _publicSharesDistributed + value_ <= _totalPublicShares,
+            "SHARE031"
+        );
+        transferFrom(owner(), to_, value_);
+        _publicSharesDistributed += value_;
+    }
+
+    /// @notice Pays out to shareholders in a round-robin fashion, where each
+    /// shareholder receives a number of payments in proportion to their balance
+    /// over a sequence of payments. The number of payments per iteration is
+    /// determined by the payment batch size parameter.
+    receive() external payable nonReentrant afterInit {
+        uint256 paymentValue = msg.value / _paymentBatchSize;
```

Comment:
TLDR: paymentBatchSize can make payments unfair, want to get clarity on my understanding or open dialogue on how to improve.

The `_paymentBatchSize` variable seems to hold significant leverage with regards to the "fairness" of this distribution system. Unlike SL2RD_v1, SL2RD_v2 is using a round robin, iterative mechanism for payment distribution. V1 uses a pseudo-random drawing system for payment distribution. 

Please correct me if I am wrong, but I would like to point out a probable scenario which highlights the leverage `_paymentBatchSize` holds to greatly skew the payment mechanism to be unfair if not held consistent. Since the pBS can be frequently(every payment) set to new sizes, unfairness would be introduced with a positive correlation between the frequency of the change and the delta of amount. It is also worthwhile to not this unfairness also increases for a non-equal distribution of shares which is a reasonable assumption for most real use cases. 
Example:
[Bob, shares: 12]<->[Alice, shares: 4]<->[Tim, shares:2]
Initialized with pBS of 2. First Payment 10 ETH, paymentValue 5 ETH.
Bob gets paid first and gets paid 10 ETH. pBS is now set to 5. Second payment is 5 ETH. Bob still has 10/12 payments owed, so he received all 5 1 ETH payments. Now pBS is set to 20 and third payment is 1 ETH. Bob is finally paid out proportional to his total shares, but Alice and Tim never had a chance to participate in this heavy positive flow. 

pBS is powerful and has no safeguards within the contract--probably for the best--, but there should either be greater callout to how to set this variable properly or a modification to the round robbin that introduces the pseudo-random drawing system based off each nodes share balance.

Reply:
@Jpsassine Thank you for the review! I think pBS does create some risk in the fairness principles here and perhaps should be immutable. That said, just thinking this through, the part "Bob gets paid first and gets paid 10 ETH"––with a pBS > 1, Bob would not get the entire 10 ETH payment but instead (10 ETH / pBS). When pBS is set to 20 and the third payment is 1 ETH, the payee pointer would increment over Alice and Tim shares since.

State chart below:

1: [Bob, shares: 12]<->[Alice, shares: 4]<->[Tim, shares:2], totalSupply = 18 shares, pbs = 2, payee node = 0

2: contract receives 10 ETH with pBS = 2. Payouts are:

- payout 1, index 0: Bob, share 0 (5ETH)
- payout 2, index 1: Bob, share 1 (5ETH)

Bob has 10ETH, Alice has 0ETH, Tim has 0ETH

3: contract receives 5 ETH with pBS = 5. Payouts are:

- payout 1, index 2: Bob, share 2 (1ETH)
- payout 2, index 3: Bob, share 3 (1ETH)
- payout 3, index 4: Bob, share 4 (1ETH)
- payout 4, index 5: Bob, share 5 (1ETH)
- payout 5, index 6: Bob, share 6 (1ETH)

4: contract receives 1 ETH with pBS = 20. Payouts are:

- payout 1, index 7: Bob, share 7 (0.05ETH)
- payout 2, index 8: Bob, share 8 (0.05ETH)
- payout 3, index 9: Bob, share 9 (0.05ETH)
- payout 4, index 10: Bob, share 10 (0.05ETH)
- payout 5, index 11: Bob, share 11 (0.05ETH)
- payout 6, index 12: Alice, share 0 (0.05ETH)
- payout 7, index 13: Alice, share 1 (0.05ETH)
- payout 8, index 14: Alice, share 2 (0.05ETH)
- payout 9, index 15: Alice, share 3 (0.05ETH)
- payout 10, index 16: Tim, share 0 (0.05ETH)
- payout 11, index 17: Tim, share 1 (0.05ETH)
- payout 12, index 18: Bob, share 0 (0.05ETH)
- payout 13, index 19: Bob, share 1 (0.05ETH)
- payout 14, index 20: Bob, share 2 (0.05ETH)
- payout 15, index 21: Bob, share 3 (0.05ETH)
- payout 16, index 22: Bob, share 4 (0.05ETH)
- payout 17, index 23: Bob, share 5 (0.05ETH)
- payout 18, index 24: Bob, share 6 (0.05ETH)
- payout 19, index 25: Bob, share 7 (0.05ETH)
- payout 20, index 26: Bob, share 8 (0.05ETH)

**Total ETH distributed: 16 ETH**
Bob receives: 15.7 ETH (98%)
Alice receives: 0.2 ETH (0.125%)
Tim receives: 0.1 ETH (0.00625%)

**Expected based on shares:**
Bob (12/18) = 66%
Alice: (4/18) = 22%
Tim: (2/18) = 11%

With this analysis I see what you're saying. Basically if you set the payment size very high and the batch size very low, early payees get paid unfairly. In general the expectation is that the payment size is constant (e.g. receiving 10 ETH, 5 ETH and 1 ETH as different payment amounts would typically not happen from upstream PFAs). I think the solution here is to publish or enforce some kind of relationship between payment amount and batch size. If the payment is 10 ETH, the batch size should be as large as possible. If the payment amount is 0.00001 ETH, the batch size should be as close to 1 as possible. I'd love to get randomization back in the algorithm but haven't figured out how to do so in this case. Since the shares are fungible, we're iterating through payees and firing off a single (with pBS == 1) payment, iteratively, based on supply of the payee, but can't iterate through individual shares ordered randomly, as shares have no order.

Reply:
Yes, your example is exactly what I was attempting to portray. Thank you!

+1, I think there needs to be an established relationship between pBS and payment size, but this should not be enforced at the contract layer. I think this would be best suited as a disclaimer in dev docs or internally while using. 

Perhaps a naive question, but why not calculate payouts per holder using something like this:

```
paymentForShareholder = (balanceOf(selectedShareholderNode.shareholderAddress) / _totalShares) * msg.value

payable(selectedShareholderNode.shareholderAddress).transfer(
                paymentForShareholder
```

and this is done inside the forloop for each shareholder, and some safeguarding to ensure msg.value is used up fully?

Reply:
> Yes, your example is exactly what I was attempting to portray. Thank you!
> 
> +1, I think there needs to be an established relationship between pBS and payment size, but this should not be enforced at the contract layer. I think this would be best suited as a disclaimer in dev docs or internally while using.
> 
> Perhaps a naive question, but why not calculate payouts per holder using something like this:
> 
> ```
> paymentForShareholder = (balanceOf(selectedShareholderNode.shareholderAddress) / _totalShares) * msg.value
> 
> payable(selectedShareholderNode.shareholderAddress).transfer(
>                 paymentForShareholder
> ```
> 
> and this is done inside the forloop for each shareholder, and some safeguarding to ensure msg.value is used up fully?

@Jpsassine Thank you! Check this out, this is how SL2RD_V2s end up looking on Uniswap. Re:

```
paymentForShareholder = (balanceOf(selectedShareholderNode.shareholderAddress) / _totalShares) * msg.value
```

The issue here is what to do with the remainder. Over time if payouts are done as fractions proportional to the supply, a remaining balance accumulates and needs to be redistributed, and from what I've been able to think through there's no way to redistribute this remainder in constant time, e.g. O(1) gas cost.

Reply:
Would the same dust problem not occur with a variable batch size, or even variable msg.values? If you truncate the fractions shouldn't this be as close to proportions while still allowing proper distribution? 

Taking a look at the on-chain activity from the link, really cool!